Move the discovery of the cluster's DNS server and subnets to the traffic-manager

thallgren opened the pull request:
## Description

Moves the discovery of the cluster's DNS server and subnets to the traffic-manager and improves the way its done. The subnet package is reintroduced (by reverting the commit that removed it) and all review comments that were made on that package were accepted and applied here instead.

The move changes the RBAC for both the client and the traffic-manager. Details in the "Adjust RBAC rules" commit.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

Commit:
Make kates.Client available as a context variable

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Revert "Remove subnet package. It's no longer needed."

This reverts commit f639e921

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fixes in response to code-review of the subnet package.

The subnet package was reviewed by @LukeShu as part of a PR from which
it later got removed. This commit accepts all the suggested changes.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Remove things from the subnet package that are not used.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make TUN-device subnet config updatable

Subnets may come and go. This commit adds the code needed to avoid
subnet duplicates and subnets covered by other subnets as the set
changes over time.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Adjust RBAC rules.

The client no longer need permission to list nodes.

The manager must be able to:
1. Create a (dummy) service in the ambassador namespace
2. List, get, and watch nodes
3. List, get, and watch pods

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Move the discovery of cluster information to the traffic-manager.

This commit moves the connector code responsible for discovering the
remote dns IP and the cluster subnets to the traffic-manager. The code
is also improved in three areas.

1. It will also look for `coredns.kube-system` when searching for the
   DNS service.
2. When we don't find podCIDRs in nodes, the pod CIDRs will be computed
   from the IPs of all pods.
3. The subnets are now dynamic and will be adjusted in runtime.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure proper close of TUN-device when context is cancelled.

Ensure proper close of the TUN-device when an error occurs in some of
its goroutines. Before this commit, that close would only happen on an
explicit quit, a fact that sometimes resulted in dangling daemon
processes.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix namespace confusion in traffic-manager manifests.

The manifests needs to be managed per namespace even if they are not
namespaced resources since the secrets involved are using DNS names
that contain the namespace.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Add service port annotation for the mutating webhook agent injector.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

donnyyung commented on CHANGELOG.md:
```diff
@@ -2,7 +2,13 @@
 ### 2.3.2 (TBD)
 
-- Bug: On a MacOS, files generated under /etc/resolver as the result of using include-suffixes in the cluster config, are now properly removed on quit.
+- Change: The list of subnets that the virtual network interface will route is configured dynamically and will follow changes in the cluster.
```
is there a docs PR w/ release notes for all these changes + any other docs changes? Or did you / will you not have time for that and need help?

donnyyung replied:
Answered offline, Thomas plans to do that tomorrow!

LukeShu commented on pkg/client/daemon/tunrouter.go:
```diff
@@ -83,6 +84,15 @@ type tunRouter struct {
+	alsoProxySubnets []*net.IPNet
+
+	// Subnets that the router is currently configured with
+	subnets []*net.IPNet
```
Would `curSubnets` be a better name?

LukeShu replied:
Perhaps also the comment should say that it should *only* be used within `refreshSubnets()`?

LukeShu commented on pkg/client/daemon/tunrouter.go:
```diff
@@ -92,37 +102,94 @@ type tunRouter struct {
+	copy(desiredSubnets[len(t.clusterSubnets):], t.alsoProxySubnets)
+	desiredSubnets = subnet.Unique(desiredSubnets)
+
+	// Trim those that are already routed
+	ln := len(desiredSubnets)
+	for i, desiredSn := range desiredSubnets {
+		if i >= ln {
+			break
+		}
+		for _, routedSn := range t.subnets {
+			if subnet.Equal(desiredSn, routedSn) {
+				ln--
+				desiredSubnets[i] = desiredSubnets[ln]
+				break
+			}
+		}
+	}
+	desiredSubnets = desiredSubnets[:ln]
```
I'm not asking you to change it, and maybe it's just me, but I found this code hard to read.  I'm not sure what I'd do about that without making it less efficient. :shrug:

donnyyung replied:
I guess maybe I'm a little confused so maybe adding a comment would be helpful...

but could you check my understanding of the inner for loop.  If the desiredSn (at index i), is already in t.subnets, then we want to "trim" it.  So we decrement the length of the desired subnets by 1 since we are going to "remove" one.  

But in  `desiredSubnets[i] = desiredSubnets[ln]`, we set the ith index to be the new decremented lnth element, but since that element is now in position `i`, will we never check if desiredSubnets[ln] is already included right?

But I see this type of loop is also used in the Unique function below, so I must just be missing something.

thallgren replied:
So "ln" is short for "length". When something that should be removed is discovered at position `i`, then that element is overwritten with the last element in the list (effectively `ln - 1` , but `ln` is decreased on the preceding line instead of after). The decrease of `ln` is necessary because the last element is now a duplicate (it's also at position `i`).

At the end of the loop, the slice is truncated using `[:ln]` which effectively removes all the duplicates.

thallgren replied:
I'll try and come up with comments that explains what goes on.

LukeShu commented on pkg/client/daemon/tunrouter.go:
```diff
@@ -92,37 +102,94 @@ type tunRouter struct {
+
+	for _, sn := range desiredSubnets {
+		if err := t.dev.AddSubnet(ctx, sn); err != nil {
+			return fmt.Errorf("failed to remove subnet %s: %w", sn, err)
```
Surely that should say "failed to add subnet".

LukeShu commented on CHANGELOG.md:
```diff
@@ -2,7 +2,13 @@
 
-- Bug: On a MacOS, files generated under /etc/resolver as the result of using include-suffixes in the cluster config, are now properly removed on quit.
+- Change: The list of subnets that the virtual network interface will route is configured dynamically and will follow changes in the cluster.
+- Change: The client no longer need permission to list nodes.
```
needs

LukeShu commented on pkg/client/connector/service_proxy.go:
```diff
@@ -188,6 +188,10 @@ func (p *mgrProxy) AgentLookupHostResponse(ctx context.Context, arg *managerrpc.
+
+func (p *mgrProxy) WatchClusterInfo(_ *managerrpc.SessionInfo, server managerrpc.Manager_WatchClusterInfoServer) error {
+	return errors.New("WatchClusterInfo is not implemented by the mgrProxy")
+}
```
I saw this and got real confused. "Won't the daemon need that!?"  I hadn't realized that the daemon hits the port-forward directly.  We should maybe be changing that.

LukeShu reviewed (changes requested):
- Change request: Typo in the changelog
- Change request: Copy/pasteod error message
- Minor suggestion: Naming and comment on `.subnets`

But on the whole :heart:

thallgren commented:
The force push contains the following:
1. Rewrite of the handling of the subnets to make the code easier to read.
2. Fixed typo on the CHANGELOG.md
3. Squashed changes into relevant commits
4. Rebased on top of release/v2 to get rid of merge conflict in changelog.

LukeShu reviewed (approved):
`Partition` is a definite readability improvement!

LukeShu commented:
It looks like CI failed because it ended up with a bum cluster? That shouldn't be happening with k3s, but whatever; I restarted it.

LukeShu commented:
Woo, CI passed, hit it!

Commit:
Extend traffic-manager API time.

This API gets hit the first time the traffic-manager boots up and
performs initial discovery of subnets so it needs to be a bit longer
to accommodate for that.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
